Add basic char*_t support for libc (partial WG14 N2653)

Febbe opened the pull request:
This PR implements a part of WG14 N2653:
 - Define C23 char8_t
 - Define C11 char16_t
 - Define C11 char32_t
 
 Missing goals are:
    - The type of UTF-8 character literals is changed from unsigned char to char8_t. (Since UTF-8 character literals already have type unsigned char, this is not a semantic change).
    - New mbrtoc8() and c8rtomb() functions declared in <uchar.h> enable conversions between multibyte characters and UTF-8.
    - A new ATOMIC_CHAR8_T_LOCK_FREE macro.
    - A new atomic_char8_t typedef name.

Febbe committed:
Add basic char*_t support for libc
 - Define C23 char8_t
 - Define C11 char16_t
 - Define C11 char32_t

Preparation for functions like `mbrtoc8` and `c8rtomb` which are
introduced in C23.

Febbe commented:
@tahonermann I am asking for a review.

Febbe committed:
check char*_t in the c23 status rst

michaelrj-google commented on libc/config/baremetal/api.td:
```diff
@@ -85,5 +85,10 @@ def TimeAPI : PublicAPI<"time.h"> {
 def UCharAPI : PublicAPI<"uchar.h"> {
-  let Types = ["mbstate_t"];
+  let Types = [
+    "mbstate_t",
```
you should also add these types to the `api.td` for `linux` if you're adding this header to the list of linux headers.

Febbe replied:
What about windows, I couldn't find any header list. Does this mean, all are included by default, or none of them?

michaelrj-google commented on libc/include/llvm-libc-types/char16_t.h:
```diff
@@ -0,0 +1,17 @@
+//===-- Definition of clock_t type ----------------------------------------===//
```
nit: this header needs to be fixed. Same for the other files.

michaelrj-google commented on libc/include/llvm-libc-types/char16_t.h:
```diff
@@ -0,0 +1,17 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIBC_TYPES_CHAR8_T_H
```
this header guard needs to be updated, same for char32

lntue commented on libc/include/llvm-libc-types/char32_t.h:
```diff
@@ -0,0 +1,17 @@
+#define LLVM_LIBC_TYPES_CHAR8_T_H
+
+#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L
+#include <stdint.h>
```
Use `#include "../llvm-libc-macros/stdint-macros.h"` and add corresponding dependency.  Similarly for `char16_t.h`.

Febbe committed:
Applied feedback
 - headers fixed
 - included `stdint-macros.h` instead of `stdint.h`
 - Updated dependencies of `char16_t` and `char32_t`
 - Added uchar support for linux-riscv
 - Added uchar & wchar support for linux-arm & linux-aarch64
 - Added UCharAPI type to linux/api.td

Febbe commented:
Thank you for the feedback. I applied all of your changes and added uchar (and wchar) also to the list of included headers on linux-aarch, linux-arm, linux-riscv

michaelrj-google commented on libc/include/llvm-libc-types/CMakeLists.txt:
```diff
@@ -90,6 +90,19 @@ add_header(tcflag_t HDR tcflag_t.h)
+  HDR 
+    char16_t.h
+  DEPENDS
+    libc.include.llvm-libc-macros.stdint_macros)
```
nit: closing parenthesis on its own line

michaelrj-google commented on libc/include/llvm-libc-types/char16_t.h:
```diff
@@ -0,0 +1,18 @@
+//===-- Definition of char16_t type
```
nit: fix the formatting on this header (you'll need to delete some `-`

michaelrj-google reviewed (commented):
Overall looks good with some minor style nits

Febbe commented:
Ok, now everything should be fixed. I can squash the commits together if that's desired. But I don't have merge rights, someone has to merge this.

michaelrj-google commented:
Github automatically squashes, I can merge this for you.
